File: 0576-out-of-boundary-paths/0576-out-of-boundary-paths.cpp
class Solution {
    void addPaths(int& res, int y)
    {
        res = (res+y)%1000000007;
    }
    int findPaths(int m, int n, int maxMove, int i, int j, vector<vector<vector<int>>>& dp)
    {
        if(i < 0 or i==m or j<0 or j ==n)
            return maxMove >= 0;
        
        if(maxMove <= 0)
            return 0;
        
        if(dp[i][j][maxMove] != -1)
            return dp[i][j][maxMove];
        
        int res = 0;
        addPaths(res, findPaths(m, n, maxMove - 1, i-1, j, dp));
        addPaths(res, findPaths(m, n, maxMove - 1, i, j+1, dp));
        addPaths(res, findPaths(m, n, maxMove - 1, i+1, j, dp));
        addPaths(res, findPaths(m, n, maxMove - 1, i, j-1, dp));     
    
        return dp[i][j][maxMove] = res;
    }
public:
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        vector<vector<vector<int>>>dp(m, vector<vector<int>>(n, vector<int>(maxMove + 1,-1)));
        return findPaths(m, n, maxMove, startRow, startColumn, dp);
    }
};

File: 0576-out-of-boundary-paths/0576-out-of-boundary-paths.py
class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        
        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        mod = 1e9 + 7
        dp = {}
        def paths(i, j, move):
                        
            if (i < 0 or i == m or j < 0 or j == n):
                return 1
            
            if move <= 0:
                return 0
            res = 0
            
            if (i, j, move) in dp:
                return dp[i, j, move]
            
            for (x, y) in directions:
                res+= paths(i + x, j + y, move - 1)
                res%=mod
            dp[i, j, move] = res    
            return res
        
        return paths(startRow, startColumn, maxMove)
            
            
            
        